### PR TITLE
WIP: Create new `Thread` preact component to replace `AnnotationThread`

### PR DESCRIPTION
**Pull request by lyzadanger**
This PR presents the proposed structure for a new preact `Thread` component to replace `AnnotationThread`.

Right now, `thread-list` renders _both_ `AnnotationThread`s and `Thread`s, so that a reviewer can compare and contrast.

I suggest pulling this branch and playing around with it, adding some nested replies, etc., and experimenting with expanding, collapsing threads as well as applying search filters. The general design approach for the threads and expand/collapse controls have been discussed and verbally approved in a [Slack conversation accessible to Hypothesis staff](https://hypothes-is.slack.com/archives/C07NXBDNW/p1585144768001900).

There is an additional design change in this set of changes for the appearance of the control to show threads hidden by applied search filters. I haven't had a chance to discuss the design of this with anyone yet. It's not meant as an end point, but it's a tad bit better than what's extant, I think.

Part of https://github.com/hypothesis/client/issues/1845

Amendment: I should point out that if you create a thread that spans more than the vertical viewport of your screen (i.e. a whole bunch of nested replies), you might end up in a state of crazy flicker and re-render as the two thread components fight with each other. This can be rectified by either removing a couple of replies from your long thread, or by commenting out the old `annotation-thread` rendering in the `thread-list` template.

### CODE REVIEW

File: src/sidebar/components/annotation.js [hunk 1]
```diff
@@ -117,15 +117,15 @@ function Annotation({
       {isEditing && <TagEditor onEditTags={onEditTags} tagList={tags} />}
       {!isEditing && <TagList annotation={annotation} tags={tags} />}
       <footer className="annotation__footer">
-        <div className="annotation__form-actions">
-          {isEditing && (
+        {isEditing && (
```
Comment (lyzadanger) [inline src/sidebar/components/annotation.js]: This seemingly unrelated change is because I realized that rendering an empty `.annotation__form-actions` `div` is part of our vertical spacing problem with annotation cards. ONE PART. MANY OTHER PARTS.

File: src/styles/sidebar/components/thread.scss [hunk 2]
```diff
@@ -0,0 +1,50 @@
+@use "../../variables" as var;
+
+.thread {
```
Comment (lyzadanger) [inline src/styles/sidebar/components/thread.scss]: This CSS is fairly prototype-quality right now—don't worry about the specifics yet, I'll clean it up.

File: src/sidebar/services/threads.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
+// @ngInject
+export default function threadsService(store) {
+  /**
+   * Make this thread and all of its children "visible". This has the effect of
+   * "unhiding" a thread which is currently hidden by an applied search filter
+   * (as well as its child threads).
+   */
+  function forceVisible(thread) {
```
Comment (lyzadanger) [inline src/sidebar/services/threads.js]: This logic reproduces what I believe several functions are trying to accomplish in `annotation-thread.js`. Start here:

https://github.com/hypothesis/client/blob/b4953ce60964ec71c10a703f6ceef2d167dad2e3/src/sidebar/components/annotation-thread.js#L60-L64

At the end of the day, this logic seems to "work." The right things are being expanded and collapsed. In this current branch, you can see both the old `annotation-thread` and the new `Thread` alongside each other, so you can test to see if any of the expand/collapse behavior is different between the two.

File: src/sidebar/components/thread.js [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+/**
+ * A thread, which contains a single annotation at its top level, and its
+ * recursively-rendered children (i.e. replies). A thread may have a parent,
+ * and at any given time it may be `collapsed`.
+ */
+function Thread({ showDocumentInfo = false, thread, threadsService }) {
+  const setCollapsed = useStore(store => store.setCollapsed);
+
+  // Only render this thread's annotation if it exists and the thread is `visible`
+  const showAnnotation = thread.annotation && thread.visible;
```
Comment (lyzadanger) [inline src/sidebar/components/thread.js]: I confess to still not being entirely clear what `visible` really means and where it gets mutated. My hunch is that `visible` threads are on-screen and, well, visible.

File: src/sidebar/components/thread.js [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+            title={toggleTitle}
+            onClick={onToggleReplies}
+          />
+        </div>
+      )}
+
+      <div className="thread__content">
+        {showAnnotation && (
+          <Fragment>
+            <ModerationBanner annotation={thread.annotation} />
```
Comment (lyzadanger) [inline src/sidebar/components/thread.js]: I think in the long run, `ModerationBanner` should be rendered within `Annotation`, not in `Thread`.

File: src/sidebar/templates/thread-list.html [hunk 6]
```diff
@@ -12,7 +12,16 @@
...
+    </div>
+  </li>
+  <li ng-repeat="child in vm.virtualThreadList.visibleThreads track by child.id">
+    <div id="{{child.id}}"
+        class="thread-list__card"
+        ng-mouseenter="vm.onFocus({annotation: child.annotation})"
+        ng-class="{'thread-list__card--theme-clean' : vm.isThemeClean }"
+        ng-click="vm.onSelect({annotation: child.annotation})"
+        ng-mouseleave="vm.onFocus({annotation: null})">
+        <thread thread="child" show-document-info="vm.showDocumentInfo"></thread>
```
Comment (lyzadanger) [inline src/sidebar/templates/thread-list.html]: There is a missing `</li>` tag here, sorry.

File: src/styles/sidebar/components/annotation-header.scss [hunk 7]
```diff
@@ -10,6 +10,7 @@
 
   &__row {
     display: flex;
+    flex-wrap: wrap-reverse;
```
Comment (lyzadanger) [inline src/styles/sidebar/components/annotation-header.scss]: This is in progress, but I need to address the situation that happens when a header (user display name plus edited timestamp plus timestamp) is too wide for its available space, which can happen 5-6 nest-levels deep sometimes. This is step one; it makes wrapping _better_ but not great yet.

File: src/styles/sidebar/components/thread.scss [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    margin-top: 0.5em;
+    padding-top: 0.5em;
+  }
+  display: flex;
+
+  &__collapse {
+    border-left: 1px dashed var.$grey-3;
+    margin: 0.25em;
+    margin-top: 0;
+    cursor: auto;
```
Comment (lyzadanger) [inline src/styles/sidebar/components/thread.scss]: I'll call this out. I did this to override the `pointer` cursor that applies across the entire annotation card. I don't want to imply that this area is clickable—only the button is.

File: src/styles/sidebar/components/thread.scss [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+  .is-collapsed &__collapse {
+    border-left: none;
+  }
+
+  &__collapse-button {
+    margin-left: -1.25em;
+    padding: 0.75em;
+    padding-top: 0.25em;
+    padding-bottom: 1em;
+    background-color: var.$white;
```
Comment (lyzadanger) [inline src/styles/sidebar/components/thread.scss]: This was necessary so that the dashed border line doesn't show through the button.

File: src/styles/sidebar/components/thread.scss [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    }
+  }
+
+  .is-collapsed &__collapse {
+    border-left: none;
+  }
+
+  &__collapse-button {
+    margin-left: -1.25em;
+    padding: 0.75em;
```
Comment (lyzadanger) [inline src/styles/sidebar/components/thread.scss]: I'll clean these up.

**Comment by robertknight** [comment]
> Right now, thread-list renders both AnnotationThreads and Threads, so that a reviewer can compare and contrast.

This confused me for a few moments ... it's cool to see this is possible now though.

File: src/sidebar/services/threads.js (hunk 3 above)
Comment (robertknight) [reply src/sidebar/services/threads.js]: This worked as I expected from a user's point of view, so the new code here is fine with me.

**Comment by robertknight** [comment]
I found an issue if you delete a reply in the middle of a chain. In this screenshot I have created an annotation and a chain of three replies at increasing depth, then attempted to delete the first reply only. The old thread renders a placeholder for the message, the new one fails to render the reply chain:

Screenshot 2020-03-26 09 39 05

Console error:

Screenshot 2020-03-26 09 41 06

File: src/sidebar/components/thread.js [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+          <Button
+            buttonText={`Show ${countHidden(thread)} more in conversation`}
+            onClick={() => threadsService.forceVisible(thread)}
+          />
+        )}
+
+        {showChildren && (
+          <ul>
+            {visibleChildren.map(child => (
+              <li key={child.annotation.$tag}>
```
Comment (robertknight) [inline src/sidebar/components/thread.js]: `child.annotation` may be null here. A `Thread` object will always have an `id` property however which you can use instead (as in `key={child.id}`).

File: src/sidebar/components/thread.js (hunk 4 above)
Comment (robertknight) [reply src/sidebar/components/thread.js]: The `Thread` data structure and the `visible` property is defined here: https://github.com/hypothesis/client/blob/d4bd1896fcf47450c36134ef60c0a0ca2234596a/src/sidebar/build-thread.js#L21

FWIW the preferred approach would be to define a JSDoc `@typedef` for a `Thread` and then reference it elsewhere to make this easier to look up.

File: src/sidebar/components/thread.js (hunk 11 above)
Comment (lyzadanger) [reply src/sidebar/components/thread.js]: Great find! Initially, I was trying to use `$tag` here to avoid exposing annotation `id`s (`thread.id` is equal to the `id` of the thread's `annotation`). But this doesn't get surfaced to the DOM rendering, so it should be fine...

Comment (lyzadanger) [reply src/sidebar/components/thread.js]: (Oh, I also fixed the alignment of the message when fixing this bug)

File: src/sidebar/services/threads.js (hunk 3 above)
Comment (lyzadanger) [reply src/sidebar/services/threads.js]: That's good news because the new variant is much simpler! Why use four functions when you can use one? :)

File: src/sidebar/components/thread.js (hunk 4 above)
Comment (lyzadanger) [reply src/sidebar/components/thread.js]: Hmmm, by the time `thread`s get to this component, they've got additional properties than those listed here, e.g. `depth` and `replyCount`. I'll do more digging.

**Comment by lyzadanger** [comment]
See cleaned up PR here: https://github.com/hypothesis/client/pull/1964
